hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ContainerEndpoint.java

```diff
@@ -498,72 +535,138 @@ private List<ContainerBlockMetadata> getBlocks(
+        });
+        break;
 
-      // Filter list of container Ids which are present in SCM but not in OM.
-      List<Long> nonOMContainers = scmNonDeletedContainers.stream()
-          .filter(containerId -> !omContainers.containsKey(containerId))
-          .collect(Collectors.toList());
+      case OM:
+        List<Long> nonOMContainers = scmNonDeletedContainers.stream()
+            .filter(containerId -> !omContainers.containsKey(containerId))
+            .collect(Collectors.toList());
 
-      List<Pipeline> pipelines = new ArrayList<>();
-      nonOMContainers.forEach(nonOMContainerId -> {
-        ContainerDiscrepancyInfo containerDiscrepancyInfo =
-            new ContainerDiscrepancyInfo();
-        containerDiscrepancyInfo.setContainerID(nonOMContainerId);
-        containerDiscrepancyInfo.setNumberOfKeys(0);
-        PipelineID pipelineID = null;
-        try {
-          pipelineID = containerManager.getContainer(
-                  ContainerID.valueOf(nonOMContainerId))
-              .getPipelineID();
-
-          if (null != pipelineID) {
-            pipelines.add(pipelineManager.getPipeline(pipelineID));
+        if (prevKey > 0) {
+          int index = 0;
+          while (index < nonOMContainers.size() &&
+              nonOMContainers.get(index) <= prevKey) {
+            index++;
+          }
+          if (index < nonOMContainers.size()) {
+            nonOMContainers = nonOMContainers.subList(index,
+                Math.min(index + limit, nonOMContainers.size()));
+          } else {
+            nonOMContainers = Collections.emptyList();
           }
-        } catch (ContainerNotFoundException e) {
-          LOG.warn("Container {} not found in SCM: {}", nonOMContainerId, e);
-        } catch (PipelineNotFoundException e) {
-          LOG.debug("Pipeline not found for container: {} and pipelineId: {}",
-              nonOMContainerId, pipelineID, e);
+        } else {
+          nonOMContainers = nonOMContainers.subList(0,
+              Math.min(limit, nonOMContainers.size()));
         }
-        containerDiscrepancyInfo.setPipelines(pipelines);
-        containerDiscrepancyInfo.setExistsAt("SCM");
-        containerDiscrepancyInfoList.add(containerDiscrepancyInfo);
-      });
 
+        List<Pipeline> pipelines = new ArrayList<>();
+        nonOMContainers.forEach(nonOMContainerId -> {
+          ContainerDiscrepancyInfo containerDiscrepancyInfo =
+              new ContainerDiscrepancyInfo();
+          containerDiscrepancyInfo.setContainerID(nonOMContainerId);
+          containerDiscrepancyInfo.setNumberOfKeys(0);
+          PipelineID pipelineID = null;
+          try {
+            pipelineID = containerManager.getContainer(
+                ContainerID.valueOf(nonOMContainerId)).getPipelineID();
+            if (pipelineID != null) {
+              pipelines.add(pipelineManager.getPipeline(pipelineID));
+            }
+          } catch (ContainerNotFoundException e) {
```

Comment:
Do we need to handle this ContainerNotFoundException, If this happens it means container is not even in SCM too. As here we are saying container not found in SCM. But we are marking in L661 as `containerDiscrepancyInfo.setExistsAt("SCM"); `.

Reply:
Thanks for the comment @ashishkumar50 The reason for throwing a `ContainerNotFoundException` and marking the container as existing in the SCM is to indicate that although the container is expected to be present in the SCM, it is not found. `containerManager.getContainer()` is called in the code that might throw a `ContainerNotFoundException` hence we would need to handle it.

Reply:
Ideally container should not be present in SCM when it is not present in OM.
Discrepancy is only when if container is present in either of SCM or OM.
But in this place when SCM throws `ContainerNotFoundException` this means that this container has no discrepancy as this container is not present even in OM.

Reply:
Thanks for the explanation @ashishkumar50 I have made changes.